camel/memory/chat_history_memory.py

```diff
@@ -0,0 +1,124 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+import pickle
+from dataclasses import asdict
+from typing import List, Optional
+
+from camel.memory.base_memory import BaseMemory
+from camel.memory.dict_storage.base import BaseDictStorage
+from camel.memory.dict_storage.in_memory import InMemoryDictStorage
+from camel.messages.base import BaseMessage
+
+
+class ChatHistoryMemory(BaseMemory):
+    """
+    An implementation of the :obj:`BaseMemory` abstract base class for
+    maintaining a record of chat histories.
+
+    This memory class helps manage conversation histories with a designated
+    storage mechanism, either provided by the user or using a default
+    in-memory storage. It offers a windowed approach to retrieving chat
+    histories, allowing users to specify how many recent messages they'd
+    like to fetch.
+
+    `ChatHistoryMemory` requires messages to be stored with certain
+    metadata (e.g., `role_at_backend`) to maintain consistency and validate
+    the chat history.
+
+    Args:
+        storage (BaseDictStorage): A storage mechanism for storing chat
+            history.
+        window_size (int, optional): Specifies the number of recent chat
+            messages to retrieve. If not provided, the entire chat history
+            will be retrieved.
+    """
+
+    def __init__(self, storage: Optional[BaseDictStorage] = None,
+                 window_size: Optional[int] = None) -> None:
+        self.storage = storage or InMemoryDictStorage()
+        self.window_size = window_size
+
+    def read(self,
+             current_state: Optional[BaseMessage] = None) -> List[BaseMessage]:
```

Comment:
Remove `current_state` temporarily since chat history memory needn't this. Considering add it in a vectorDB PR. Maybe later I will come up with a better design.